### PR TITLE
docstrings improvements for composer.datasets

### PR DESCRIPTION
**Pull request by dskhudia**
For __init__, ade20k, brats, c4, cifar10, coco, dataloader, dataset_registry and imagenet.

### CODE REVIEW

File: composer/datasets/__init__.py [hunk 1]
```diff
@@ -1,32 +1,42 @@
...
+                                         SyntheticDataType, SyntheticPILDataset)
+
+__all__ = [
+    "ADE20kDatasetHparams", "ADE20kWebDatasetHparams", "BratsDatasetHparams", "C4DatasetHparams",
+    "CIFAR10DatasetHparams", "CIFAR10WebDatasetHparams", "CIFAR20WebDatasetHparams", "CIFAR100WebDatasetHparams",
+    "COCODatasetHparams", "DataloaderHparams", "WrappedDataLoader", "get_dataset_registry", "EvaluatorHparams",
+    "GLUEHparams", "DatasetHparams", "SyntheticHparamsMixin", "WebDatasetHparams", "Imagenet1kWebDatasetHparams",
+    "ImagenetDatasetHparams", "TinyImagenet200WebDatasetHparams", "LMDatasetHparams", "MNISTDatasetHparams",
+    "MNISTWebDatasetHparams", "MemoryFormat", "SyntheticBatchPairDataset", "SyntheticDataLabelType",
+    "SyntheticDataType", "SyntheticPILDataset"
```
Comment (hanlint) [inline composer/datasets/__init__.py:41]: For these, and also the above, can we do:
```
__all__ = [
1,
2,
3,
4,
]
```

format?

File: composer/datasets/dataloader.py [hunk 2]
```diff
@@ -143,7 +155,7 @@ def initialize_object(
     ) -> DataLoader:
         """Create a dataloader.
 
-        Args:
+        Parameters:
```
Comment (hanlint) [inline composer/datasets/dataloader.py:158]: Args?

File: composer/datasets/imagenet.py [hunk 3]
```diff
@@ -1,5 +1,11 @@
 # Copyright 2021 MosaicML. All Rights Reserved.
 
+"""ImageNet classfication dataset.
+
+The most widely used dataset for Image Classification algorithms. For more details about this dataset, please refer to
+`ImageNet 2012 Classification Dataset <http://image-net.org/>`_.
```
Comment (hanlint) [inline composer/datasets/imagenet.py:6]:
```suggestion
`ImageNet 2012 Classification Dataset <http://image-net.org/>`_.

Also includes streaming dataset versions based on `WebDatasets <https://github.com/webdataset/webdataset>`_.
```

File: composer/datasets/__init__.py [hunk 4]
```diff
@@ -1,32 +1,42 @@
...
-from composer.datasets.mnist import MNISTDatasetHparams as MNISTDatasetHparams
-from composer.datasets.mnist import MNISTWebDatasetHparams as MNISTWebDatasetHparams
-from composer.datasets.synthetic import MemoryFormat as MemoryFormat
-from composer.datasets.synthetic import SyntheticBatchPairDataset as SyntheticBatchPairDataset
-from composer.datasets.synthetic import SyntheticDataLabelType as SyntheticDataLabelType
-from composer.datasets.synthetic import SyntheticDataType as SyntheticDataType
-from composer.datasets.synthetic import SyntheticPILDataset as SyntheticPILDataset
+* All datasets derive from the abstract base class :class:`~.DatasetHparams` and it contains common parameters such
+
+  or a :class:`~.DataSpec`) used by the trainer.
```
Comment (growlix) [inline composer/datasets/__init__.py]: is there text that should have gone between "...common parameters such" and "or a :class:..."?

File: composer/datasets/__init__.py [hunk 5]
```diff
@@ -1,32 +1,42 @@
...
+
+* :class:`~.DataloaderHparams` contains the :class:`torch.utils.data.Dataloader` settings that are common across
```
Comment (growlix) [inline composer/datasets/__init__.py:12]: Link for
```
:class:`torch.utils.data.Dataloader`
```
not rendering

File: composer/datasets/ade20k.py [hunk 6]
```diff
@@ -1,5 +1,11 @@
 # Copyright 2021 MosaicML. All Rights Reserved.
 
+"""Semantic segmentation and scene parsing dataset.
+
+For more details about this dataset, please refere to `ADE20k dataset
```
Comment (growlix) [inline composer/datasets/ade20k.py]:
```suggestion
For more details about this dataset, please refer to `ADE20k dataset
```

File: composer/datasets/ade20k.py [hunk 7]
```diff
@@ -20,9 +26,11 @@
...
 
+__all__ = ["ADE20k", "ADE20kDatasetHparams", "ADE20kWebDatasetHparams"]
+
 
 class RandomResizePair(torch.nn.Module):
-    """Resize the image and target to `base_size` scaled by a randomly sampled value.
+    """Resize the image and target to ``base_size`` scaled by a randomly sampled value.
 
     Args:
         min_scale (float): the minimum value the samples can be rescaled.
```
Comment (growlix) [inline composer/datasets/ade20k.py:36]: base_size (and many parameters in this file) do not have a default specified or need the default formatted correctly. Defaults should be specified with the following formatting:
```
Default: ``value``.
```

File: composer/datasets/ade20k.py [hunk 8]
```diff
@@ -241,7 +249,7 @@ def __len__(self):
 
 @dataclass
 class ADE20kDatasetHparams(DatasetHparams, SyntheticHparamsMixin):
-    """Defines an instance of the ADE20k dataset for semantic segmentation.
+    """Defines an instance of the ADE20k dataset for semantic segmentation from a local disk.
 
     Args:
         split (str): the dataset split to use either 'train', 'val', or 'test'. Default is `train`.
```
Comment (growlix) [inline composer/datasets/ade20k.py:255]:
```suggestion
        split (str): the dataset split to use either 'train', 'val', or 'test'. Default: ``'train'``.
```

File: composer/datasets/__init__.py (hunk 1 above)
Comment (dskhudia) [reply composer/datasets/__init__.py:41]: I tried it the suggested way but make style re-formatted it back.

File: composer/datasets/c4.py [hunk 9]
```diff
@@ -140,7 +145,7 @@ class C4Dataset(IterableDataset):
         shuffle_buffer_size (int): If `shuffle=True`, samples are read into a buffer of this size (per-device), and randomly sampled from there to produce shuffled samples. (Default: `10000`)
```
Comment (growlix) [inline composer/datasets/c4.py:145]: The line above seems too long. Try running `make style` to see if this gets fixed.

File: composer/datasets/c4.py [hunk 10]
```diff
@@ -130,7 +135,7 @@ class C4Dataset(IterableDataset):
     HuggingFace's C4 Dataset with streaming backend (See https://huggingface.co/datasets/c4 for more details). The text
     samples are then shuffled, tokenized, and grouped on-the-fly.
 
-    Parameters:
+    Args:
```
Comment (growlix) [inline composer/datasets/c4.py:138]: Default formatting

File: composer/datasets/dataloader.py [hunk 11]
```diff
@@ -91,9 +103,9 @@ def unwrap_data_loader(dataloader: DataLoader) -> DataLoader:
 
 @dataclass
 class DataloaderHparams(hp.Hparams):
-    """Hyperparameters to initialize a :class:`~torch.utils.data.Dataloader`.
+    """Hyperparameters to initialize a :class:`torch.utils.data.Dataloader`.
```
Comment (growlix) [inline composer/datasets/dataloader.py:106]: Class link is not rendering

**Review by growlix — COMMENTED** [review]
Looks pretty good! Mostly just need to format Args defaults (though it seems like there may not be [consensus on the formatting ATM](https://mosaicml.slack.com/archives/C02F05H39SL/p1646790497668119))

File: composer/datasets/__init__.py (hunk 4 above)
Comment (dskhudia) [reply composer/datasets/__init__.py]: yes. Somehow it got removed. Fixed it.

File: composer/datasets/__init__.py (hunk 5 above)
Comment (dskhudia) [reply composer/datasets/__init__.py:12]: Fixed it.

File: composer/datasets/c4.py [hunk 12]
```diff
@@ -38,22 +43,27 @@ def _split_dict_fn(batch: Batch, n_microbatches: int) -> List[Batch]:
...
+        max_seq_len (int): The max sequence length of each token sample. Default: ``None``.
         group_method (str): How to group text samples into token samples. Either `truncate` or `concat`.
-        mlm (bool): Whether or not to use masked language modeling. (Default: `False`)
-        mlm_probability (float): If `mlm=True`, the probability that tokens are masked. (Default: `0.15`)
-        shuffle (bool): Whether to shuffle the samples in the dataset. Currently, shards are assigned and consumed with deterministic per-device shard order, but shuffling affects the order of samples via (per-device) shuffle buffers. (Default: `False`)
-        shuffle_buffer_size (int): If `shuffle=True`, samples are read into a buffer of this size (per-device), and randomly sampled from there to produce shuffled samples. (Default: `10000`)
-        seed (int): If `shuffle=True`, what seed to use for shuffling operations. (Default: `5`)
-        drop_last (bool): Whether to drop the last samples for the last batch. (Default: `True`)
+            Default: ``None``.
+        mlm (bool): Whether or not to use masked language modeling. Default: ``False``
```
Comment (growlix) [inline composer/datasets/c4.py:56]:
```suggestion
        mlm (bool): Whether or not to use masked language modeling. Default: ``False``.
```

File: composer/datasets/c4.py [hunk 13]
```diff
@@ -38,22 +43,27 @@ def _split_dict_fn(batch: Batch, n_microbatches: int) -> List[Batch]:
...
+        mlm_probability (float): If ``mlm=True``, the probability that tokens are masked. Default: ``0.15``.
+        shuffle (bool): Whether to shuffle the samples in the dataset. Currently, shards are assigned and consumed with
+            deterministic per-device shard order, but shuffling affects the order of samples via (per-device) shuffle
+            buffers. Default: ``False``
```
Comment (growlix) [inline composer/datasets/c4.py:60]:
```suggestion
            buffers. Default: ``False``.
```

File: composer/datasets/c4.py [hunk 14]
```diff
@@ -38,22 +43,27 @@ def _split_dict_fn(batch: Batch, n_microbatches: int) -> List[Batch]:
...
+        shuffle_buffer_size (int): If ``shuffle=True``, samples are read into a buffer of this size (per-device), and
+            randomly sampled from there to produce shuffled samples. Default: ``10000``
```
Comment (growlix) [inline composer/datasets/c4.py:62]:
```suggestion
            randomly sampled from there to produce shuffled samples. Default: ``10000``.
```

File: composer/datasets/c4.py [hunk 15]
```diff
@@ -38,22 +43,27 @@ def _split_dict_fn(batch: Batch, n_microbatches: int) -> List[Batch]:
...
+        seed (int): If ``shuffle=True``, what seed to use for shuffling operations. Default: ``5``
```
Comment (growlix) [inline composer/datasets/c4.py:63]:
```suggestion
        seed (int): If ``shuffle=True``, what seed to use for shuffling operations. Default: ``5``.
```

File: composer/datasets/c4.py [hunk 16]
```diff
@@ -38,22 +43,27 @@ def _split_dict_fn(batch: Batch, n_microbatches: int) -> List[Batch]:
...
+        drop_last (bool): Whether to drop the last samples for the last batch. Default: ``True``
```
Comment (growlix) [inline composer/datasets/c4.py:64]:
```suggestion
        drop_last (bool): Whether to drop the last samples for the last batch. Default: ``True``.
```

File: composer/datasets/c4.py [hunk 17]
```diff
@@ -38,22 +43,27 @@ def _split_dict_fn(batch: Batch, n_microbatches: int) -> List[Batch]:
...
     Returns:
-        A :class:`~composer.core.DataSpec` object
+        DataSpec: A :class:`.DataSpec` object
```
Comment (growlix) [inline composer/datasets/c4.py:66]:
```suggestion
        DataSpec: A :class:`.DataSpec` object.
```

File: composer/datasets/c4.py [hunk 18]
```diff
@@ -130,17 +140,20 @@ class C4Dataset(IterableDataset):
...
         max_seq_len (int): The max sequence length of each token sample.
-        group_method (str): How to group text samples into token samples. Either `truncate` or `concat`.
-        shuffle (bool): Whether to shuffle the samples in the dataset. Currently, shards are assigned and consumed with deterministic per-device shard order, but shuffling affects the order of samples via (per-device) shuffle buffers. (Default: `False`)
-        shuffle_buffer_size (int): If `shuffle=True`, samples are read into a buffer of this size (per-device), and randomly sampled from there to produce shuffled samples. (Default: `10000`)
-        seed (int): If `shuffle=True`, what seed to use for shuffling operations. (Default: `5`)
+        group_method (str): How to group text samples into token samples. Either ``'truncate'`` or ``'concat'``.
+        shuffle (bool): Whether to shuffle the samples in the dataset. Currently, shards are assigned and consumed with
+            deterministic per-device shard order, but shuffling affects the order of samples via (per-device) shuffle
+            buffers. Default: ``False``
```
Comment (growlix) [inline composer/datasets/c4.py:151]:
```suggestion
            buffers. Default: ``False``.
```

File: composer/datasets/c4.py [hunk 19]
```diff
@@ -130,17 +140,20 @@ class C4Dataset(IterableDataset):
...
+        shuffle_buffer_size (int): If ``shuffle=True``, samples are read into a buffer of this size (per-device), and
+            randomly sampled from there to produce shuffled samples. Default: ``10000``
```
Comment (growlix) [inline composer/datasets/c4.py:153]:
```suggestion
            randomly sampled from there to produce shuffled samples. Default: ``10000``.
```

File: composer/datasets/c4.py [hunk 20]
```diff
@@ -130,17 +140,20 @@ class C4Dataset(IterableDataset):
...
+        seed (int): If ``shuffle=True``, what seed to use for shuffling operations. Default: ``5``
```
Comment (growlix) [inline composer/datasets/c4.py:154]:
```suggestion
        seed (int): If ``shuffle=True``, what seed to use for shuffling operations. Default: ``5``.
```

File: composer/datasets/c4.py [hunk 21]
```diff
@@ -130,17 +140,20 @@ class C4Dataset(IterableDataset):
...
     Returns:
-        A :class:`torch.utils.data.IterableDataset` object
+        IterableDataset: A :class:`torch.utils.data.IterableDataset` object
```
Comment (growlix) [inline composer/datasets/c4.py:156]:
```suggestion
        IterableDataset: A :class:`torch.utils.data.IterableDataset` object.
```

File: composer/datasets/dataloader.py [hunk 22]
```diff
@@ -91,23 +103,23 @@ def unwrap_data_loader(dataloader: DataLoader) -> DataLoader:
...
-    """Hyperparameters to initialize a :class:`~torch.utils.data.Dataloader`.
+    """Hyperparameters to initialize a :class:`torch.utils.data.DataLoader`.
 
-    Parameters:
+    Args:
         num_workers (int, optional): Number of CPU workers to use per device to fetch data.
             Set to ``0`` to use the main training thread for dataloading.
             While zero workers can be useful for debugging, it should not be used for performance reasons.
-            (default: ``8``)
+            Default: ``8``
```
Comment (growlix) [inline composer/datasets/dataloader.py:112]:
```suggestion
            Default: ``8``.
```

File: composer/datasets/dataloader.py [hunk 23]
```diff
@@ -91,23 +103,23 @@ def unwrap_data_loader(dataloader: DataLoader) -> DataLoader:
...
         prefetch_factor (int, optional): Number of samples loaded in advance by each worker.
             For example, 2 means there will be a total of 2 * num_workers samples prefetched across all workers.
             If ``num_workers = 0``, then the ``prefetch_factor`` must be left at the default value.
-            (default: ``2``)
+            Default: ``2``
```
Comment (growlix) [inline composer/datasets/dataloader.py:116]:
```suggestion
            Default: ``2``.
```

File: composer/datasets/dataloader.py [hunk 24]
```diff
@@ -91,23 +103,23 @@ def unwrap_data_loader(dataloader: DataLoader) -> DataLoader:
...
         persistent_workers (bool): Whether to reuse dataloader workers across epochs. If ``num_workers`` is 0,
-            then this field must be ``False``. (default: ``True``)
+            then this field must be ``False``. Default: ``True``
```
Comment (growlix) [inline composer/datasets/dataloader.py:118]:
```suggestion
            then this field must be ``False``. Default: ``True``.
```

File: composer/datasets/dataloader.py [hunk 25]
```diff
@@ -91,23 +103,23 @@ def unwrap_data_loader(dataloader: DataLoader) -> DataLoader:
...
         pin_memory (bool, optional): Whether or not to copy Tensors into CUDA pinned memory before returning them.
-            If ``num_workers = 0``, then the ``pin_memory`` must be ``False``. (default: ``True``)
+            If ``num_workers = 0``, then the ``pin_memory`` must be ``False``. Default: ``True``
```
Comment (growlix) [inline composer/datasets/dataloader.py:120]:
```suggestion
            If ``num_workers = 0``, then the ``pin_memory`` must be ``False``. Default: ``True``.
```

File: composer/datasets/dataloader.py [hunk 26]
```diff
@@ -91,23 +103,23 @@ def unwrap_data_loader(dataloader: DataLoader) -> DataLoader:
...
         timeout (float): Timeout, in seconds, for collecting a batch from workers. Set to ``0`` for no timeout.
-            (default: ``0``)
+            Default: ``0``
```
Comment (growlix) [inline composer/datasets/dataloader.py:122]:
```suggestion
            Default: ``0``.
```

**Review by growlix — COMMENTED** [review]
A few small formatting things but otherwise looks great, Daya!

**Comment by growlix** [comment]
lgtm!

**Review by growlix — APPROVED** [review]
LGTM!

**Pull request merged**